File: applier/src/Applier.java
import android.content.Context;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.telephony.CarrierConfigManager;
import android.telephony.SubscriptionInfo;
import android.telephony.SubscriptionManager;
import android.telephony.TelephonyManager;
import org.json.JSONArray;
import org.json.JSONObject;
import java.lang.reflect.Method;
import java.util.List;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;

/**
 * Root-mode CarrierConfig applier. Runs via
 *   CLASSPATH=Applier.dex app_process / Applier <action>
 * as uid 0 (root). Root holds MODIFY_PHONE_STATE implicitly, so
 * CarrierConfigManager.overrideConfig succeeds without FLAG_SYSTEM,
 * a priv-app, or any /system overlay — zero boot-loop risk.
 *
 * Logic ported from the original ImsModifier.buildBundle (Shizuku app).
 */
public class Applier {
    static final String CONFIG_PATH = "/data/adb/carrier_ims/config.json";
    static final String STATUS_PATH = "/data/adb/carrier_ims/status.json";

    static final String K_NR_THR_BW = "nr_advanced_threshold_bandwidth_khz_int";
    static final String K_NR_ADV_BANDS = "additional_nr_advanced_bands_int_array";
    static final String K_5G_ICON = "5g_icon_configuration_string";
    static final String K_NR_PCO = "nr_advanced_capable_pco_id_int";
    static final String K_NR_INC_LTE = "include_lte_for_nr_advanced_threshold_bandwidth_bool";
    static final int NR_THR_5GA = 110000;
    static final String NR_ICON_5GA =
        "connected_mmwave:5G_Plus,connected:5G,connected_rrc_idle:5G,not_restricted_rrc_idle:5G,not_restricted_rrc_con:5G";
    static final int[] NR_BANDS_CN = {1,3,8,28,41,78,79};
    static final String K_MCC_OVERRIDE = "country_mcc_override";
    static final String MCC_CN = "460";

    public static void main(String[] args) throws Throwable {
        String action = (args.length > 0) ? args[0] : "apply";

        // app_process launched from shell has no Application. Use systemMain()
        // to create a system-level ActivityThread and get a usable Context.
        Class<?> at = Class.forName("android.app.ActivityThread");
        Object thread = at.getMethod("systemMain").invoke(null);
        Context ctx = (Context) at.getMethod("getSystemContext").invoke(thread);
        if (ctx == null) { writeError("cannot get system context"); return; }
        CarrierConfigManager cm = (CarrierConfigManager) ctx.getSystemService(Context.CARRIER_CONFIG_SERVICE);
        SubscriptionManager sm = (SubscriptionManager) ctx.getSystemService(Context.TELEPHONY_SUBSCRIPTION_SERVICE);
        TelephonyManager tm = (TelephonyManager) ctx.getSystemService(Context.TELEPHONY_SERVICE);

        if (cm == null || sm == null) { writeError("system service unavailable"); return; }

        @SuppressWarnings("unchecked")
        List<SubscriptionInfo> subs = (List<SubscriptionInfo>)
            SubscriptionManager.class.getMethod("getActiveSubscriptionInfoList").invoke(sm);
        if (subs == null || subs.isEmpty()) { writeError("no active subscriptions"); return; }

        if ("reset".equals(action)) {
            for (SubscriptionInfo sub : subs) invoke(cm, sub.getSubscriptionId(), null, false);
            writeReset(); return;
        }

        File cfgFile = new File(CONFIG_PATH);
        if (!cfgFile.exists()) { writeError("no config"); return; }
        JSONObject cfg = new JSONObject(readFile(cfgFile));
        JSONObject slots = cfg.optJSONObject("slots");
        if (slots == null) { writeError("no slots configured"); return; }

        JSONArray results = new JSONArray();
        for (SubscriptionInfo sub : subs) {
            int slot = sub.getSimSlotIndex();
            int subId = sub.getSubscriptionId();
            JSONObject sc = slots.optJSONObject(String.valueOf(slot));
            if (sc == null) continue;

            Bundle b = buildBundle(sc);
            boolean applied = false; String error = null;
            try { invoke(cm, subId, toPB(b), true); applied = true; }
            catch (Throwable pe) {
                try { invoke(cm, subId, toPB(b), false); applied = true; }
                catch (Throwable fe) { error = fe.getMessage() != null ? fe.getMessage() : fe.getClass().getSimpleName(); }
            }
            boolean ims = imsRegistered(tm, subId);
            JSONObject r = new JSONObject();
            r.put("slotIndex", slot); r.put("subId", subId);
            r.put("applied", applied); r.put("imsRegistered", ims);
            if (error != null) r.put("error", error);
            results.put(r);
        }
        JSONObject status = new JSONObject();
        status.put("lastApplyMillis", System.currentTimeMillis());
        status.put("slots", results);
        writeFile(STATUS_PATH, status.toString());
    }

    static Bundle buildBundle(JSONObject s) throws Exception {
        Bundle b = new Bundle();
        if (s.optBoolean("tiktokNetworkFix", false)) {
            b.putString(CarrierConfigManager.KEY_SIM_COUNTRY_ISO_OVERRIDE_STRING, "cn");
            b.putString(K_MCC_OVERRIDE, MCC_CN);
        }
        if (s.optBoolean("volte", true)) {
            b.putBoolean(CarrierConfigManager.KEY_CARRIER_VOLTE_AVAILABLE_BOOL, true);
            b.putBoolean(CarrierConfigManager.KEY_EDITABLE_ENHANCED_4G_LTE_BOOL, true);
            b.putBoolean(CarrierConfigManager.KEY_HIDE_ENHANCED_4G_LTE_BOOL, false);
            b.putBoolean(CarrierConfigManager.KEY_HIDE_LTE_PLUS_DATA_ICON_BOOL, false);
        }
        if (s.optBoolean("show4gForLte", false)) b.putBoolean("show_4g_for_lte_data_icon_bool", true);
        if (s.optBoolean("vt", true)) b.putBoolean(CarrierConfigManager.KEY_CARRIER_VT_AVAILABLE_BOOL, true);
        if (s.optBoolean("ut", true)) b.putBoolean(CarrierConfigManager.KEY_CARRIER_SUPPORTS_SS_OVER_UT_BOOL, true);
        if (s.optBoolean("crossSim", true)) {
            b.putBoolean(CarrierConfigManager.KEY_CARRIER_CROSS_SIM_IMS_AVAILABLE_BOOL, true);
            b.putBoolean(CarrierConfigManager.KEY_ENABLE_CROSS_SIM_CALLING_ON_OPPORTUNISTIC_DATA_BOOL, true);
        }
        if (s.optBoolean("vowifi", true)) {
            b.putBoolean(CarrierConfigManager.KEY_CARRIER_WFC_IMS_AVAILABLE_BOOL, true);
            b.putBoolean(CarrierConfigManager.KEY_CARRIER_WFC_SUPPORTS_WIFI_ONLY_BOOL, true);
            b.putBoolean(CarrierConfigManager.KEY_EDITABLE_WFC_MODE_BOOL, true);
            b.putBoolean(CarrierConfigManager.KEY_EDITABLE_WFC_ROAMING_MODE_BOOL, true);
            b.putBoolean("show_wifi_calling_icon_in_status_bar_bool", true);
            b.putInt("wfc_spn_format_idx_int", 6);
        }
        if (s.optBoolean("vonr", true)) {
            b.putBoolean(CarrierConfigManager.KEY_VONR_ENABLED_BOOL, true);
            b.putBoolean(CarrierConfigManager.KEY_VONR_SETTING_VISIBILITY_BOOL, true);
        }
        if (s.optBoolean("fiveGnr", true)) {
            b.putIntArray(CarrierConfigManager.KEY_CARRIER_NR_AVAILABILITIES_INT_ARRAY,
                new int[]{CarrierConfigManager.CARRIER_NR_AVAILABILITY_NSA, CarrierConfigManager.CARRIER_NR_AVAILABILITY_SA});
            if (s.optBoolean("fiveGPlusIcon", true)) {
                b.putInt(K_NR_THR_BW, NR_THR_5GA);
                b.putBoolean(K_NR_INC_LTE, false);
                b.putIntArray(K_NR_ADV_BANDS, NR_BANDS_CN);
                b.putString(K_5G_ICON, NR_ICON_5GA);
                b.putInt(K_NR_PCO, 0);
            }
            if (s.optBoolean("fiveGThresholds", true)) {
                b.putIntArray(CarrierConfigManager.KEY_5G_NR_SSRSRP_THRESHOLDS_INT_ARRAY,
                    new int[]{-128,-118,-108,-98});
            }
        }
        return b;
    }

    static PersistableBundle toPB(Bundle b) {
        PersistableBundle pb = new PersistableBundle();
        for (String k : b.keySet()) {
            Object v = b.get(k);
            if (v instanceof Boolean) pb.putBoolean(k, (Boolean)v);
            else if (v instanceof Integer) pb.putInt(k, (Integer)v);
            else if (v instanceof String) pb.putString(k, (String)v);
            else if (v instanceof int[]) pb.putIntArray(k, (int[])v);
        }
        return pb;
    }

    static void invoke(CarrierConfigManager cm, int subId, PersistableBundle v, boolean persistent) throws Throwable {
        try {
            cm.getClass().getMethod("overrideConfig", int.class, PersistableBundle.class, boolean.class)
                .invoke(cm, subId, v, persistent);
        } catch (NoSuchMethodException e) {
            cm.getClass().getMethod("overrideConfig", int.class, PersistableBundle.class)
                .invoke(cm, subId, v);
        }
    }

    static boolean imsRegistered(TelephonyManager tm, int subId) {
        try {
            TelephonyManager per = (TelephonyManager)
                TelephonyManager.class.getMethod("createForSubscriptionId", int.class).invoke(tm, subId);
            Object r = TelephonyManager.class.getMethod("isImsRegistered", int.class).invoke(per, subId);
            return r != null && (Boolean) r;
        } catch (Throwable t) { return false; }
    }

    static String readFile(File f) throws Exception {
        byte[] buf = new byte[(int)f.length()];
        FileInputStream fis = new FileInputStream(f);
        fis.read(buf); fis.close();
        return new String(buf, "UTF-8");
    }
    static void writeFile(String path, String content) throws Exception {
        FileWriter w = new FileWriter(path); w.write(content); w.close();
    }
    static void writeError(String msg) {
        try {
            JSONObject s = new JSONObject();
            s.put("lastApplyMillis", System.currentTimeMillis());
            s.put("error", msg); s.put("slots", new JSONArray());
            writeFile(STATUS_PATH, s.toString());
        } catch (Exception ignored) {}
    }
    static void writeReset() {
        try {
            JSONObject s = new JSONObject();
            s.put("lastApplyMillis", System.currentTimeMillis());
            s.put("reset", true); s.put("slots", new JSONArray());
            writeFile(STATUS_PATH, s.toString());
        } catch (Exception ignored) {}
    }
}

File: privapp/src/main/java/io/carrierims/applier/Applier.kt
package io.carrierims.applier

import android.content.Context
import android.os.PersistableBundle
import android.telephony.CarrierConfigManager
import android.telephony.SubscriptionInfo
import android.telephony.SubscriptionManager
import android.telephony.TelephonyManager
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Applies CarrierConfig overrides for every active subscription whose
 * slotIndex is present in [ConfigStore.ModuleConfig.slots].
 *
 * Persistence strategy (solves "每次启动重新执行"):
 *  - Primary: overrideConfig(subId, bundle, persistent=true). This succeeds
 *    because the app is installed under /system/priv-app/ (→ FLAG_SYSTEM) and
 *    holds MODIFY_PHONE_STATE (signature|privileged) via the privapp-permissions
 *    XML — the exact gate that ShizukuProvider.kt:168 enforces.
 *  - Contingency: if a ROM still rejects persistent, fall back to non-persistent
 *    overrideConfig; the boot + SIM-change auto re-apply covers the gap.
 */
object Applier {

    private const val TAG = "Applier"

    data class SlotResult(
        val slotIndex: Int,
        val subId: Int,
        val applied: Boolean,
        val imsRegistered: Boolean,
        val error: String? = null,
    )

    data class ApplyOutcome(
        val lastApplyMillis: Long,
        val enabled: Boolean,
        val slots: List<SlotResult>,
    )

    fun apply(context: Context, config: ConfigStore.ModuleConfig): ApplyOutcome {
        val results = mutableListOf<SlotResult>()
        val now = System.currentTimeMillis()

        if (!config.enabled) {
            writeStatus(now, enabled = false, emptyList())
            return ApplyOutcome(now, enabled = false, slots = emptyList())
        }

        val cm = context.getSystemService(CarrierConfigManager::class.java)
        val sm = context.getSystemService(SubscriptionManager::class.java)
        val tm = context.getSystemService(TelephonyManager::class.java)

        val activeSubs: List<SubscriptionInfo> = try {
            @Suppress("DEPRECATION")
            sm?.activeSubscriptionInfoList ?: emptyList()
        } catch (t: Throwable) {
            Log.e(TAG, "getActiveSubscriptionInfoList failed", t)
            emptyList()
        }

        if (cm == null || sm == null || tm == null) {
            results.add(
                SlotResult(
                    slotIndex = -1,
                    subId = -1,
                    applied = false,
                    imsRegistered = false,
                    error = "system service unavailable",
                ),
            )
            writeStatus(now, enabled = true, results)
            return ApplyOutcome(now, enabled = true, slots = results)
        }

        for (sub in activeSubs) {
            val slotIndex = sub.simSlotIndex
            val subId = sub.subscriptionId
            val slotConfig = config.slots[slotIndex]
            if (slotConfig == null) {
                Log.i(TAG, "no config for slot $slotIndex (subId $subId), skip")
                continue
            }
            val bundle = ConfigBuilder.build(slotConfig)
            val applied = try {
                invokeOverrideConfig(cm, subId, bundle, persistent = true)
                Log.i(TAG, "overrideConfig persistent success for subId $subId (slot $slotIndex)")
                true
            } catch (persistentError: Throwable) {
                Log.w(TAG, "persistent failed for subId $subId, fallback non-persistent", persistentError)
                try {
                    invokeOverrideConfig(cm, subId, bundle, persistent = false)
                    Log.i(TAG, "fallback non-persistent success for subId $subId")
                    true
                } catch (fallbackError: Throwable) {
                    fallbackError.addSuppressed(persistentError)
                    Log.e(TAG, "overrideConfig failed for subId $subId", fallbackError)
                    false
                }
            }

            val imsRegistered = try {
                // createForSubscriptionId(int) and isImsRegistered(int) are
                // hidden @SystemApi; reflect to compile against the public SDK.
                val perSubTm = TelephonyManager::class.java
                    .getMethod("createForSubscriptionId", Int::class.javaPrimitiveType)
                    .invoke(tm, subId) as? TelephonyManager
                perSubTm?.let {
                    TelephonyManager::class.java
                        .getMethod("isImsRegistered", Int::class.javaPrimitiveType)
                        .invoke(it, subId) as? Boolean
                } ?: false
            } catch (t: Throwable) {
                Log.w(TAG, "isImsRegistered read failed for subId $subId", t)
                false
            }

            results.add(
                SlotResult(
                    slotIndex = slotIndex,
                    subId = subId,
                    applied = applied,
                    imsRegistered = imsRegistered,
                    error = if (applied) null else "overrideConfig rejected",
                ),
            )
        }

        writeStatus(now, enabled = true, results)
        return ApplyOutcome(now, enabled = true, slots = results)
    }

    /** Reflective overrideConfig: try 3-arg (subId, bundle, persistent) then 2-arg. */
    private fun invokeOverrideConfig(
        cm: CarrierConfigManager,
        subId: Int,
        values: PersistableBundle?,
        persistent: Boolean,
    ) {
        try {
            cm.javaClass.getMethod(
                "overrideConfig",
                Int::class.javaPrimitiveType,
                PersistableBundle::class.java,
                Boolean::class.javaPrimitiveType,
            ).invoke(cm, subId, values, persistent)
        } catch (_: NoSuchMethodException) {
            cm.javaClass.getMethod(
                "overrideConfig",
                Int::class.javaPrimitiveType,
                PersistableBundle::class.java,
            ).invoke(cm, subId, values)
        }
    }

    private fun writeStatus(now: Long, enabled: Boolean, slots: List<SlotResult>) {
        try {
            val arr = JSONArray()
            for (s in slots) {
                arr.put(
                    JSONObject()
                        .put("slotIndex", s.slotIndex)
                        .put("subId", s.subId)
                        .put("applied", s.applied)
                        .put("imsRegistered", s.imsRegistered)
                        .also { o -> s.error?.let { o.put("error", it) } },
                )
            }
            val status = JSONObject()
                .put("lastApplyMillis", now)
                .put("enabled", enabled)
                .put("slots", arr)
            File(ConfigStore.CONFIG_DIR).mkdirs()
            File(ConfigStore.STATUS_PATH).writeText(status.toString())
            Log.i(TAG, "status.json written")
        } catch (t: Throwable) {
            Log.e(TAG, "write status.json failed", t)
        }
    }
}
